runtime/src/accounts_db.rs

```diff
@@ -1432,35 +1856,59 @@ impl AccountsDB {
     /// Scan a specific slot through all the account storage in parallel
     pub fn scan_account_storage<F, B>(&self, slot: Slot, scan_func: F) -> Vec<B>
     where
-        F: Fn(&StoredAccount, AppendVecId, &mut B) + Send + Sync,
+        F: Fn(LoadedAccount, AppendVecId, &mut B) + Send + Sync,
         B: Send + Default,
     {
         self.scan_account_storage_inner(slot, scan_func)
     }
 
     fn scan_account_storage_inner<F, B>(&self, slot: Slot, scan_func: F) -> Vec<B>
     where
-        F: Fn(&StoredAccount, AppendVecId, &mut B) + Send + Sync,
+        F: Fn(LoadedAccount, AppendVecId, &mut B) + Send + Sync,
         B: Send + Default,
     {
-        let storage_maps: Vec<Arc<AccountStorageEntry>> = self
-            .storage
-            .get_slot_stores(slot)
-            .map(|res| res.read().unwrap().values().cloned().collect())
-            .unwrap_or_default();
-        self.thread_pool.install(|| {
-            storage_maps
-                .into_par_iter()
-                .map(|storage| {
-                    let accounts = storage.accounts.accounts(0);
-                    let mut retval = B::default();
-                    accounts.into_iter().for_each(|stored_account| {
-                        scan_func(&stored_account, storage.append_vec_id(), &mut retval)
-                    });
-                    retval
-                })
-                .collect()
-        })
+        if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
+            // If we see the slot in the cache, then all the account information
+            // is in this cached slot
+            let mut retval = B::default();
+            for cached_account in slot_cache.iter() {
+                scan_func(
+                    LoadedAccount::Cached((
+                        *cached_account.key(),
+                        Cow::Borrowed(cached_account.value()),
+                    )),
+                    CACHE_VIRTUAL_STORAGE_ID,
+                    &mut retval,
+                );
+            }
+            vec![retval]
```

Comment:
@ryoqun @sakridge, so the scan here across the `DashMap` cache no longer uses the `par_iter()` that the `AccountStorageEntry` uses below, because `DashMap` does not implement the `IntoParallelRefIterator` trait. In theory this means that where before the scans of the accounts were sharded across many storage entries that could be loaded in parallel, we are now sequentially scanning the accounts in the cache.

We could try to implement `IntoParallelRefIterator`, but even Rayon's implementation of `IntoParallelRefIterator`  for std collections like `HashMap` first collects into a vector, https://docs.rs/rayon/1.5.0/src/rayon/collections/mod.rs.html#12-20, and as all the callers of `scan_account_storage()` right now, most importantly `get_accounts_delta_hash()` only do a single scan to clone out the `pubkey` and `hash`, implementing `IntoParallelRefIterator`  in a similar fashion as Rayon does for the std collections probably won't yield much of a performance benefit, so we'll have to make a custom threading solution to load from a `DashMap` in parallel (maybe get inner access to the different inner hashmaps).

Moreover, the perf for `get_accounts_delta_hash` (not done yet, but we can now additionally remove the `merge()` step in that function) after this change, at least for now, seems to have improved, so a custom implementation does not seem needed yet:

<img width="1711" alt="Screen Shot 2021-01-01 at 7 55 01 PM" src="https://user-images.githubusercontent.com/3374799/103450613-f6299a80-4c6d-11eb-8f13-438c71cf28d0.png">

Reply:
Edit: https://docs.rs/dashmap/4.0.1/src/dashmap/rayon/map.rs.html#185 😄